src/utils.cpp

```diff
@@ -109,3 +130,163 @@ KeyRingReceive ZypperUtils::keyReport()
+	const ResolverProblemList& problems( s_zypp->resolver()->problems() );
+	cout << problems.size() << " problems found..." << endl;
+	
+	ProblemSolutionList toTry;
+	unsigned int probNo = 0;
+	for (const auto& probPtr : problems) {
+	    cout << "Problem " << ++probNo << ": " << probPtr->description() << endl;
+	    const ProblemSolutionList & solutions = probPtr->solutions();
+	    unsigned int solNo = 0;
+	    for ( const auto & solPtr : solutions ) {
+		if (solNo == 0) {
+		  solNo++;
+		  continue;
+		}
+		cout << "  Solution " << solNo << ": " << solPtr->description() << endl;
+		//toTry.push_back( solPtr);
+	    }
+	}
+	
+	if ( !toTry.empty() )
+	{
+	    cout << "Apply selected solutions..." << endl;
+	    s_zypp->resolver()->applySolutions( toTry );
+	    cout << "Solving dependencies..." << endl;
+	    continue;
+	}
+	// otherwise give up
+      throw "Solving dependencies failed: Giving up!";
+    }
+    cout << "Dependencies solved" << endl; */
+    
+    return s_zypp->resolver()->resolvePool();
+}
+
+bool ZypperUtils::commitChanges()
+{
+    cout << "Selectable summary (grouped by name):" << endl;
+    for ( const ui::Selectable_Ptr & sel : s_zypp->pool().proxy() )
+    {
+	if ( sel->toModify() )
+	cout << "  " << sel << endl;
+    }
+    
+    // Commit the changes 
+    // Please note that dryRunFlag and zypp::DownloadOnly are for now
+    cout << "Committing the changes" << endl;
+    bool dryRunFlag = false;
+    ZYppCommitPolicy policy;
+    if ( !dryRunFlag ) {
+	policy.dryRun( true );
+	dryRunFlag = true;
+    }
+    policy.downloadMode( zypp::DownloadOnly );
+    
+    try {
+	ZYppCommitResult result = s_zypp->commit( policy );
+	if ( !( result.allDone() || (dryRunFlag && result.noError() ) ) )
+	    throw "Incomplete Commit";
+	cout << "Commit Succeeded" << endl;
+    }
+    catch ( const Exception& exp ) {
+	cout << "Commit aborted with exception" << endl;
+	throw;	
+    }
+}
+
+void ZypperUtils::initTarget( const Pathname& sysRoot )
+{
+    cout << "Initializing target at " << sysRoot << endl;
+    s_zypp->initializeTarget( sysRoot ); //initialize target
+    cout << "Loading target resolvables" << endl;
+    s_zypp->getTarget()->load(); // load installed packages to pool
+}
+
+void ZypperUtils::makeNecessaryChangesToRunAsRoot( const Pathname& sysRoot )
+{
+    /* Caller: ZypperUtils::install()
+     * Can be absolutely certain that s_repoManagerOpts, and s_repoManager
+     * are already intialized [to display information]
+     */
+    try {
+	delete s_repoManagerOpts;
+	delete s_repoManager;
+    }
+    catch ( const Exception& exp ) {
+	cout << "Oops! You're trying to delete the non-existent. ;)" << endl;
+    }
+
+    // Initialize s_repoManager with sysRoot. This step is necessary to add repo
+    // to system repos
+    try {
+	s_repoManager = new RepoManager( sysRoot );
+    }
+    catch ( const Exception& exp ) {
+	cout << "Allocating memory to s_repoManager Failed" << endl;
+    }
+}
```

Comment:
Sorry, but I don't get the intention behind this code. :)

C++ laws are:
 - a DTOR must not throw
 - `delete null_ptr` is safe
So the `try { delete...` should never catch anything (and even if it could, a throw in the 1st delete would create a leak as the 2nd is not executed...).

Furthermore you don't reset `s_repoManagerOpts` and `s_repoManager`  If now also the `new RepoManager`  would fail, you'd leave the method  with both global pointer being not null and still pointing to meanwhile deleted memory (s_repoManagerOpts will do this anyway). 

Also if `new RepoManager` throws an `Exception` (a zypp::Exception), then the object has been allocated but the RepoManager CTOR threw an exception. The message to stdout is wrong, the `s_repoManager` pointer is iinvalid and you nevertheless leave the function? Consider re-throwing exceptions if you report, but can't handle them. 
IMO a crash is better than undefined behavior and life is too short to waste time in a debugger ;)

IMHO one should never use `delete`; and `new` only to immediately wrap the object into a smart pointer. A scoped_ptr, shared_ptr,.. everything is better than manually managing dynamically allocated resources (new/delete, malloc/free, open/close, ...), especially in code using exceptions. Use RAII to handle resources.
```
// Define static variables here
// RepoManagerOptions* ZypperUtils::s_repoManagerOpts should not be needed at all
scoped_ptr<RepoManager> ZypperUtils::s_repoManager( new RepoManager( "/tmp ") );
...
{
  s_repoManager.reset();
  try {
    s_repoManager.reset( new RepoManager( sysRoot ) );
  }
  catch ( const Exception& exp ) {
    cerr << "Oops! failed to init repo manager." << endl;
    throw;
  }
```